mindmeld/auto_annotator.py

```diff
@@ -835,5 +839,300 @@ def valid_entity_check(self, entity):
+                override the config specified by the app's config.py file.
+        """
+        super().__init__(app_path=app_path, config=config)
+        self.language = self.config.get("language", "en")
+        self.translator = self.get_translator()
+        self.en_annotator = SpacyAnnotator(app_path)
+
+    def get_translator(self):
+        """Creates a Translator object based on the translator class specified in the config.
+        If a translator is not specified, Duckling heuristics will be used.
+
+        Returns:
+            translator (Translator): Translator specified in the config, None if not specified.
+        """
+        if self.config.get("translator"):
+            translator_class_name = self.config.get("translator", "GoogleTranslator")
+            if translator_class_name == "GoogleTranslator":
+                return GoogleTranslator()
+            else:
+                logger.warning(
+                    "%s is not a valid translation class. Currently only ‘GoogleTranslator’"
+                    " is supported.", translator_class_name
+                )
+        logger.info(
+            "No Translator selected. MLA will detect entities based on Duckling heuristics."
+        )
+
+    def parse(self, sentence, entity_types=None, language=None, **kwargs):
+        """
+        Args:
+            sentence (str): Sentence to detect entities.
+            entity_types (list): List of entity types to parse. If None, all
+                    possible entity types will be parsed.
+            language (str): Language code for the language of the given sentence
+        Returns: entities (list): List of entity dictionaries.
+        """
+        language = language or self.language
+        params = {"sentence": sentence, "language": language}
+        duckling_entities = (
+            self._parse_with_translator(**params)
+            if self.translator
+            else self._parse_without_translator(**params)
+        )
+        if entity_types:
+            duckling_entities = [
+                e for e in duckling_entities if e["dim"] in entity_types
+            ]
+        return MultiLingualAnnotator.convert_duckling_candidates(duckling_entities)
+
+    def _parse_with_translator(self, sentence, language=None):
+        """Parse helper function to be used if a translator is present.
+
+        Args:
+            sentence (str): Sentence to detect entities.
+            language (str): Language code for the language of the given sentence
+        Returns: entities (list): List of duckling candidates.
+        """
+        candidates = self.en_annotator.duckling.get_candidates_for_text(
+            sentence, language=language
+        )
+        en_sentence = self.translator.translate(sentence, target_language="en")
+        en_entities = self.en_annotator.parse(en_sentence)
+
+        selected_candidates = []
+        for entity in en_entities:
+            i = 0
+            while i < len(candidates):
```

Comment:
Why not `for candidate in candidates`? It seems like you break when a match is found?

Reply:
With the above, you could disregard `i`

Reply:
Good catch, switched to `for candidate in candidates`.